File: android/src/main/java/io/invertase/firebase/crashlytics/RNFirebaseCrashlytics.java
package io.invertase.firebase.crashlytics;

import android.util.Log;
import android.os.Handler;

// import com.crashlytics.android.Crashlytics;
import com.google.firebase.crashlytics.FirebaseCrashlytics;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.ReadableArray;
import com.facebook.react.bridge.ReadableMap;

import java.util.ArrayList;

// import io.fabric.sdk.android.Fabric;

public class RNFirebaseCrashlytics extends ReactContextBaseJavaModule {

  private static final String TAG = "RNFirebaseCrashlytics";

  public RNFirebaseCrashlytics(ReactApplicationContext reactContext) {
    super(reactContext);
    Log.d(TAG, "New instance");
  }

  @Override
  public String getName() {
    return TAG;
  }

  @ReactMethod
  public void crash() {
    // throw new RuntimeException("Test Crash");
    new Handler().postDelayed(new Runnable() {
        @Override
        public void run() {
          throw new RuntimeException("Crash Test");
        }
      }, 50);
  }

  @ReactMethod
  public void log(final String message) {
    FirebaseCrashlytics.getInstance().log(message);
  }

  @ReactMethod
  public void recordError(final int code, final String domain) {
    FirebaseCrashlytics.getInstance().recordException(new Exception(code + ": " + domain));
  }

  @ReactMethod
  public void recordCustomError(String name, String reason, ReadableArray frameArray) {
      ArrayList<StackTraceElement> stackList = new ArrayList<>(0);
      for (int i = 0; i < frameArray.size(); i++) {
        ReadableMap map = frameArray.getMap(i);
        ReadableMap additional = map.hasKey("additional") ? map.getMap("additional") : null;
        String functionName = map.hasKey("functionName") ? map.getString("functionName") : "Unknown Function";
        String className = map.hasKey("className") ? map.getString("className") : "Unknown Class";
        StackTraceElement stack = new StackTraceElement(
          className,
          functionName,
          map.getString("fileName"),
          map.hasKey("lineNumber") ? map.getInt("lineNumber") : -1
        );
        stackList.add(stack);

        if(additional != null){
          StackTraceElement s = new StackTraceElement(
            "Additional Parameters",
            additional.toString(),
            map.getString("fileName"),
            map.hasKey("lineNumber") ? map.getInt("lineNumber") : -1
          );
          stackList.add(s);
        }
      }
      StackTraceElement[] stackTrace =  new StackTraceElement[stackList.size()];
      Exception e = new Exception(name + "\n" + reason);
      stackTrace = stackList.toArray(stackTrace);
      e.setStackTrace(stackTrace);
      FirebaseCrashlytics.getInstance().recordException(e);
  }

  // @ReactMethod
  // public void setBoolValue(final String key, final boolean boolValue) {
  //   FirebaseCrashlytics.getInstance().setCustomKey(key, boolValue);
  // }

  @ReactMethod
  public void setFloatValue(final String key, final float floatValue) {
    FirebaseCrashlytics.getInstance().setCustomKey(key, floatValue);
  }

  @ReactMethod
  public void setIntValue(final String key, final int intValue) {
    FirebaseCrashlytics.getInstance().setCustomKey(key, intValue);
  }

  @ReactMethod
  public void setStringValue(final String key, final String stringValue) {
    FirebaseCrashlytics.getInstance().setCustomKey(key, stringValue);
  }

  @ReactMethod
  public void setUserIdentifier(String userId) {
    FirebaseCrashlytics.getInstance().setUserId(userId);
  }

  // @ReactMethod
  // public void setUserName(String userName) {
  //   FirebaseCrashlytics.getInstance().setUserName(userName);
  // }

  // @ReactMethod
  // public void setUserEmail(String userEmail) {
  //   FirebaseCrashlytics.getInstance().setUserEmail(userEmail);
  // }

  // @ReactMethod
  // public void enableCrashlyticsCollection() {
  //   Fabric.with(getReactApplicationContext(), new Crashlytics());
  // }

  @ReactMethod
  public void enableCrashlyticsCollection(final boolean boolValue) {
    FirebaseCrashlytics.getInstance().setCrashlyticsCollectionEnabled(boolValue);
  }

}

File: ios/RNFirebase/fabric/crashlytics/RNFirebaseCrashlytics.m
#import "RNFirebaseCrashlytics.h"

// #if __has_include(<Crashlytics/Crashlytics/Crashlytics.h>)
//     #define HAS_CRASHLYTICS 1
//     #import <Crashlytics/Crashlytics/Crashlytics.h>
// #else
//     #if __has_include(<Crashlytics/Crashlytics.h>)
//     #define HAS_CRASHLYTICS 2
//     #import <Crashlytics/Crashlytics.h>
//     #endif
// #endif

// #if __has_include(<Fabric/Fabric.h>)
//     #import <Fabric/Fabric.h>
// #endif

#import <Firebase.h>
#define HAS_CRASHLYTICS 1

#ifdef HAS_CRASHLYTICS
    @implementation RNFirebaseCrashlytics
    RCT_EXPORT_MODULE();

    RCT_EXPORT_METHOD(crash) {
        @throw NSInternalInconsistencyException;
    }

    RCT_EXPORT_METHOD(log:(NSString *)message) {
        [[FIRCrashlytics crashlytics] log:@"first_arg, second_arg"];
    }

    RCT_EXPORT_METHOD(recordError:(nonnull NSNumber *)code domain:(NSString *)domain) {
        NSError *error = [NSError errorWithDomain:domain code:[code integerValue] userInfo:nil];
        [[FIRCrashlytics crashlytics] recordError:error];
    }

//    RCT_EXPORT_METHOD(recordCustomError:(nonnull NSString *)name reason:(NSString *)reason frameArray:(nonnull NSArray *)frameArray)
//    {
//        NSMutableArray *clsFrames = [[NSMutableArray alloc] init];
//        if(frameArray) {
//            for (NSDictionary *dict in frameArray) {
//            CLSStackFrame *frame = [CLSStackFrame stackFrame];
//            [frame setLibrary: dict[@"className"]];
//            [frame setFileName: dict[@"fileName"]];
//            [frame setLineNumber: [dict[@"lineNumber"] intValue]];
//            //[frame setOffset: [dict[@"columnNumber"] intValue]]; //Add later?
//            [frame setSymbol: dict[@"functionName"]];
//            [clsFrames addObject: frame];
//
//            if(dict[@"additional"]){
//                CLSStackFrame *f = [CLSStackFrame stackFrame];
//                [f setLibrary: @"Additional Parameters"];
//                [f setSymbol: [dict[@"additional"] stringValue]];
//                [f setFileName: dict[@"fileName"]];
//                [f setLineNumber: [dict[@"lineNumber"] intValue]];
//                [clsFrames addObject: f];
//            }
//
//            }
//            [[Crashlytics sharedInstance] recordCustomExceptionName:name reason:reason frameArray:clsFrames];
//        }
//    }

//    RCT_EXPORT_METHOD(setBoolValue:(NSString *)key boolValue:(BOOL *)boolValue) {
//        [[FIRCrashlytics crashlytics] setCustomValue:boolValue forKey:key];
//    }

    RCT_EXPORT_METHOD(setFloatValue:(NSString *)key floatValue:(nonnull NSNumber *)floatValue) {
        [[FIRCrashlytics crashlytics] setCustomValue:floatValue forKey:key];
    }

    RCT_EXPORT_METHOD(setIntValue:(NSString *)key intValue:(nonnull NSNumber *)intValue) {
        [[FIRCrashlytics crashlytics] setCustomValue:intValue forKey:key];
    }

    RCT_EXPORT_METHOD(setStringValue:(NSString *)key stringValue:(NSString *)stringValue) {
        [[FIRCrashlytics crashlytics] setCustomValue:stringValue forKey:key];
    }

    RCT_EXPORT_METHOD(setUserIdentifier:(NSString *)userId) {
        [[FIRCrashlytics crashlytics] setUserID:userId];
    }

    RCT_EXPORT_METHOD(enableCrashlyticsCollection:(BOOL *)boolValue) {
        [[FIRCrashlytics crashlytics] setCrashlyticsCollectionEnabled:boolValue];
    }

//    RCT_EXPORT_METHOD(setUserName:(NSString *)userName)
//    {
//        [[Crashlytics sharedInstance] setUserName:userName];
//    }
//
//    RCT_EXPORT_METHOD(setUserEmail:(NSString *)email)
//    {
//        [[Crashlytics sharedInstance] setUserEmail:email];
//    }

//    RCT_EXPORT_METHOD(enableCrashlyticsCollection) {
//#if __has_include(<Fabric/Fabric.h>)
//        [Fabric with:@[[Crashlytics class]]];
//#endif
//    }

    @end
#else
    @implementation RNFirebaseCrashlytics
    @end
#endif
